[WIP] LLMQs Implementation

codablock opened the pull request:
This is my current state of the LLMQ implementation. It's for review only and not meant for merging for now. It also contains https://github.com/dashpay/dash/pull/2296. There are also still a few things to do, also in regard to compatibility to the actual DIPs. Some things are not implemented yet and for other things it turned out that the DIPs need to be modified.

If you want to experiment with it, there is the `./qa/rpc-tests/quorums.py` script available which will locally start up a few MNs and leave them running until you kill that script. It will generate the necessary blocks to fund the MNs and activate DIP3. After all MNs have been started, sporks are put in place so that you can start testing LLMQs. At this point, no LLMQs have been created yet.

Regtest has a single LLMQ type configured, which creates a LLMQ of size 10 every 24 blocks. Give the DKG enough time (a second per block) when you generate blocks by yourself.

After a LLMQ has been activated/mined, send some TXs around to see LLMQ signing happen.

Connect to the MN network with either RPC or (as I do) a dash-qt instance with this configuration:
```
datadir=<mydatadir>
regtest=1

# needed if you want to control sporks from dash-qt
sporkkey=cP4EKFyJsHT39LDqgdcB43Y3YXjNyjb5Fuas1GQSeAtjnZWmZEQK

# listen as this dash-qt instance will also act as a MN
listen=1
bind=0.0.0.0
port=31001
externalip=127.0.0.1

allowprivatenet=1

# This one might be different on your machine, check out the running MNs and their configs (somewhere in /tmp) to find the correct port
addnode=127.0.0.1:12001

# Activates the budget system faster (probably not needed, but then it must be removed from quorums.py as well)
budgetparams=240:100:240

# Let's watch all quorums
watchquorums=1

# Let's run dash-qt as masternode
masternode=1

# probably not needed for these tests? We immediately activate DIP3 which then requires BLS keys
masternodeprivkey=cUP3knDebUNdFDRNAivB1Qcw4SQscmQgv7hUTYnuiwk39r3wDieE

######
# chia BLS12-381 (reversed)
#masternodeblsprivkey=3b7724d56e02f7948db20bc85e8b72c2fa4e0febd2d8b9315200835f1f54e5cf

```

PastaPastaPasta commented on src/crypto/bls_worker.cpp:
```diff
@@ -0,0 +1,989 @@
+        }
+        if (vvec->size() != vvecs[start]->size()) {
+            return false;
+        }
```
could combine those using `||`

PastaPastaPasta commented on src/crypto/bls_worker.cpp:
```diff
@@ -0,0 +1,989 @@
+                           CBLSWorker::SignDoneCallback doneCallback)
+{
+    workerPool.push([secKey, msgHash, doneCallback](int threadId) {
+       doneCallback(secKey.Sign(msgHash));
```
Shouldn't be three spaces

PastaPastaPasta commented on src/crypto/bls_worker.cpp:
```diff
@@ -0,0 +1,989 @@
+    std::unique_ptr<SigVerifyJob> job;
+    SigVerifyJob* jobPtr;
+
+    struct Batch {
```
bracket position

PastaPastaPasta commented on src/crypto/bls_worker.cpp:
```diff
@@ -0,0 +1,989 @@
+            });
+        }
+    }
+}
```
\n

PastaPastaPasta commented on src/ctpl.h:
```diff
@@ -0,0 +1,240 @@
+            return f;
+        }
+
+
```
extra line, rm imo

codablock replied:
ctpl.h is a header only library, so not our code. Resolving comments for this file...

PastaPastaPasta commented on src/ctpl.h:
```diff
@@ -0,0 +1,240 @@
+}
+
+#endif // __ctpl_thread_pool_H__
+
```
extra line

PastaPastaPasta commented on src/evo/providertx.cpp:
```diff
@@ -42,15 +42,31 @@ static bool CheckService(const uint256& proTxHash, const ProTx& proTx, const CBl
+{
+    std::string strError;
+    if (!CHashSigner::VerifyHash(::SerializeHash(proTx), keyID, proTx.vchSig, strError))
+        return state.DoS(100, false, REJECT_INVALID, "bad-protx-sig", false, strError);
```
body same line or brackets

PastaPastaPasta commented on src/evo/providertx.cpp:
```diff
@@ -42,15 +42,31 @@ static bool CheckService(const uint256& proTxHash, const ProTx& proTx, const CBl
+static bool CheckSig(const ProTx& proTx, const CBLSPublicKey& pubKey, CValidationState& state)
+{
+    if (!proTx.sig.Verify(pubKey, CBLSMessageHash(::SerializeHash(proTx), pubKey)))
+        return state.DoS(100, false, REJECT_INVALID, "bad-protx-sig", false);
```
same

PastaPastaPasta commented on src/evo/providertx.cpp:
```diff
@@ -42,15 +42,31 @@ static bool CheckService(const uint256& proTxHash, const ProTx& proTx, const CBl
-    if (!CHashSigner::VerifyHash(::SerializeHash(proTx), keyID, proTx.vchSig, strError))
-        return state.DoS(100, false, REJECT_INVALID, "bad-protx-sig", false, strError);
+    if (!CheckSig(proTx, pubKey, state))
+        return false;
```
same

PastaPastaPasta commented on src/evo/specialtx.cpp:
```diff
@@ -93,6 +101,9 @@ bool ProcessSpecialTxsInBlock(const CBlock& block, const CBlockIndex* pindex, CV
         return false;
 
+    if (!llmq::quorumBlockProcessor->ProcessBlock(block, pindex->pprev, state))
+        return false;
```
same line or brackets

PastaPastaPasta commented on src/evo/specialtx.cpp:
```diff
@@ -104,6 +115,9 @@ bool UndoSpecialTxsInBlock(const CBlock& block, const CBlockIndex* pindex)
     }
 
+    if (!llmq::quorumBlockProcessor->UndoBlock(block, pindex))
+        return false;
```
same

PastaPastaPasta commented on src/governance-vote.cpp:
```diff
@@ -249,7 +269,21 @@ bool CGovernanceVote::IsValid(bool useVotingKey) const
+    if (useVotingKey) {
+        CheckSignature(infoMn.keyIDVoting);
+    } else {
+        if (deterministicMNManager->IsDeterministicMNsSporkActive()) {
```
combine to single `else if {`

PastaPastaPasta commented on src/init.cpp:
```diff
@@ -825,6 +831,9 @@ bool InitSanityCheck(void)
         return false;
 
+    if (!BLSInit())
+        return false;
```
same line or brackets

PastaPastaPasta commented on src/llmq/quorums.cpp:
```diff
@@ -0,0 +1,469 @@
+#include "quorums_dkg.h"
+#include "quorums_dkgmgr.h"
+#include "quorums_utils.h"
+#include "quorums_blockprocessor.h"
```
alphabetize

PastaPastaPasta commented on src/llmq/quorums.cpp:
```diff
@@ -0,0 +1,469 @@
+#include "validation.h"
+#include "univalue.h"
+#include "chainparams.h"
+#include "activemasternode.h"
```
alphabetize

PastaPastaPasta commented on src/llmq/quorums.h:
```diff
@@ -0,0 +1,170 @@
+#define DASH_QUORUMS_H
+
+#include "evo/evodb.h"
+#include "evo/deterministicmns.h"
```
alphabetize

PastaPastaPasta commented on src/llmq/quorums.h:
```diff
@@ -0,0 +1,170 @@
+#include "evo/deterministicmns.h"
+
+#include "validationinterface.h"
+#include "consensus/params.h"
```
alphabetize

PastaPastaPasta commented on src/llmq/quorums_dkg.cpp:
```diff
@@ -0,0 +1,1169 @@
+#include "netmessagemaker.h"
+#include "univalue.h"
+#include "chainparams.h"
+#include "spork.h"
```
alphabetize

PastaPastaPasta commented on src/llmq/quorums_dkg.cpp:
```diff
@@ -0,0 +1,1169 @@
+
+#include "quorums_dkg.h"
+#include "quorums_dkgmgr.h"
+#include "quorums.h"
```
alphabetize
```
#include "quorums.h"
#include "quorums_dkg.h"
#include "quorums_dkgmgr.h"
```

PastaPastaPasta commented on src/llmq/quorums_dkg.cpp:
```diff
@@ -0,0 +1,1169 @@
+    const uint64_t v = 100000000;
+    uint64_t r = GetRand(v + 1);
+    if (r <= v * rate)
+        return true;
```
same line or brackets

PastaPastaPasta commented on src/llmq/quorums_dkg.cpp:
```diff
@@ -0,0 +1,1169 @@
+    return !seenMessages.emplace(msgHash).second;
+}
+
+}
```
\n

PastaPastaPasta commented on src/llmq/quorums_dkg.h:
```diff
@@ -0,0 +1,323 @@
+
+#include "consensus/params.h"
+#include "net.h"
+#include "batchedlogger.h"
```
alphabetize

PastaPastaPasta commented on src/llmq/quorums_dkgmgr.cpp:
```diff
@@ -0,0 +1,639 @@
+void CDKGType::WaitForNextPhase(QuorumPhase curPhase,
+                                                QuorumPhase nextPhase,
+                                                uint256& expectedBlockHash,
+                                                const WhileWaitFunc& runWhileWaiting)
```
spacing

PastaPastaPasta commented on src/llmq/quorums_dkgmgr.cpp:
```diff
@@ -0,0 +1,639 @@
+void CDKGType::RandomSleep(QuorumPhase curPhase,
+                                           uint256& expectedBlockHash,
+                                           double randomSleepFactor,
+                                           const WhileWaitFunc& runWhileWaiting)
```
spacing

PastaPastaPasta commented on src/llmq/quorums_dkgmgr.cpp:
```diff
@@ -0,0 +1,639 @@
+                                           uint256& expectedBlockHash,
+                                           double randomSleepFactor,
+                                           const StartPhaseFunc& startPhaseFunc,
+                                           const WhileWaitFunc& runWhileWaiting)
```
spacing

PastaPastaPasta commented on src/llmq/quorums_dkgmgr.cpp:
```diff
@@ -0,0 +1,639 @@
+    return true;
+}
+
+////
```
rm

PastaPastaPasta commented on src/llmq/quorums_dkgmgr.cpp:
```diff
@@ -0,0 +1,639 @@
+bool CDKGManager::GetPrematureCommitment(const uint256& hash, CDKGPrematureCommitment& ret) const
+{
+    if (!sporkManager.IsSporkActive(SPORK_16_QUORUM_DKG_ENABLED))
+        return false;
```
same line or bracket

PastaPastaPasta commented on src/llmq/quorums_dkgmgr.cpp:
```diff
@@ -0,0 +1,639 @@
+bool CDKGManager::GetFinalCommitment(const uint256& hash, CFinalCommitment& ret) const
+{
+    if (!sporkManager.IsSporkActive(SPORK_16_QUORUM_DKG_ENABLED))
+        return false;
```
same line or bracket

PastaPastaPasta commented on src/net_processing.cpp:
```diff
@@ -44,6 +44,10 @@
+#include "llmq/quorums.h"
+#include "llmq/quorums_dkgmgr.h"
+#include "llmq/quorums_signing.h"
+#include "llmq/quorums_instantx.h"
```
alphabetize

PastaPastaPasta commented on src/privatesend.h:
```diff
@@ -12,6 +12,7 @@
 #include "sync.h"
 #include "tinyformat.h"
 #include "timedata.h"
+#include "crypto/bls.h"
```
alphabetize

PastaPastaPasta commented on src/rpc/governance.cpp:
```diff
@@ -393,20 +401,8 @@ UniValue gobject_vote_conf(const JSONRPCRequest& request)
-
     CGovernanceVote vote(mn.outpoint, hash, eVoteSignal, eVoteOutcome);
-    if (!vote.Sign(activeMasternodeInfo.keyOperator, activeMasternodeInfo.keyIDOperator)) {
+    if(!vote.Sign(activeMasternodeInfo.legacyKeyOperator, activeMasternodeInfo.legacyKeyIDOperator)) {
```
`if(` -> `if (`

PastaPastaPasta commented on src/rpc/rpcquorums.cpp:
```diff
@@ -0,0 +1,179 @@
+
+#include "server.h"
+#include "base58.h"
+#include "validation.h"
```
alphabetize

PastaPastaPasta commented on src/rpc/rpcquorums.cpp:
```diff
@@ -0,0 +1,179 @@
+UniValue quorum_list(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 1 && request.params.size() != 2)
+        quorum_list_help();
```
same line or brackets

PastaPastaPasta commented on src/rpc/rpcquorums.cpp:
```diff
@@ -0,0 +1,179 @@
+UniValue quorum_info(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 3 && request.params.size() != 4)
+        quorum_info_help();
```
same line or brackets

PastaPastaPasta commented on src/test/evo_deterministicmns_tests.cpp:
```diff
@@ -392,3 +392,4 @@ BOOST_FIXTURE_TEST_CASE(dip3_protx, TestChainDIP3Setup)
     BOOST_ASSERT(foundRevived);
 }
 BOOST_AUTO_TEST_SUITE_END()
+*/
```
\n

PastaPastaPasta commented on src/util.h:
```diff
@@ -244,6 +244,11 @@ int GetNumCores();
 void RenameThread(const char* name);
 std::string GetThreadName();
 
+namespace ctpl {
```
brace on newline for namespaces

PastaPastaPasta commented on src/bench/bls.cpp:
```diff
@@ -0,0 +1,347 @@
+#include "bench.h"
+#include "random.h"
+#include "crypto/bls_worker.h"
+#include "utiltime.h"
```
alphabetize

codablock commented:
Just force-pushed my current version. This version is now based on the current develop branch (v13).

A lot has changed in the code, especially the way I propagate and validate signature shares between LLMQ members. In the previous version, I was directly pushing sigshare batches to intra-quorum connected nodes, without checking if these already have these shares. I did this because the classical INV system would already send around messages of at least 32 bytes size, so I assumed directly sending the shares itself would actually reduce the overhead of the INV system. Turned out it still produces too much overhead and chatter between nodes, especially as we now have switched to BLS12-381 with 96 byte signatures. With this switch, even the classical INV system would probably be better performing then what I did initially (direct push).

Instead of switching the the INV system, I implemented a specialized inventory system for signature shares. It should perform MUCH better and with much less overhead then the general purpose INV system.

Instead of announcing INV items to the other nodes, LLMQ members will now send `CSigSharesInv` messages to other members. It contains some information to identify the signing session to which the shares belong and a bitset that represents the shares that to offer. Each bit represents a share from a LLMQ member. Bit at index 10 means it's a share from member 10.

Members will now regularly (every 100ms) check which shares were announced by other members and which shares are missing locally and request these from other members. They will also check what other nodes requested and send them these shares. All the communication is leveraging the `CSigSharesInv` object to reduce overhead to the minimum and to request/announce as many shares as possible in one go. Shares are still sent with the `CBatchedSigShares` to reduce overhead even further.

Also, in the old version I extensively used the `CBLSWorker` to speed up things in the handling of signature shares. I removed this and will also remove the use of it in other places later. Turned out that the overhead due to moving BLS primitives around between threads destroys most of the performance gains. Especially now as we switched to BLS12-381, where the primitives got a lot larger.

Instead, I'm leveraging (single-threaded) batched verification as much as possible now. Incoming shares are put into a pending batch and then processed every 100ms. Verification is then done in sub-batches to avoid DoS attacks (an attacker sending 100 valid shares and one last invalid one). I first try to batch together from all nodes and if that succeeds, all shares are accepted. If it fails, I revert to per-node batching to figure out which node sent the invalid shares.

I'll document all this in-code when I find time, as reviewing it will otherwise not be fun.

Another change is how I determine to which nodes to announce shares to. Previously, I only sent shares to members which were determined by the intra-quorum connections selection algorithm. This however had the disadvantage that it only announced shares to outgoing connections and did not take incoming connection from other members into account. As there is no easy way to determine which incoming connection is from a LLMQ member, I implemented something like "if someone ever sends me a share belonging to my quorum, he must be interested in my shares as well, so I announce him all the shares I got from now on". It doesn't necessarily mean it's really a member of the same LLMQ, but its 99% likely.

codablock commented on src/base58.cpp:
```diff
@@ -251,9 +261,15 @@ bool CBitcoinAddress::IsValid() const
+                         vchVersion == params.Base58Prefix(CChainParams::SCRIPT_ADDRESS) ||
+                         vchVersion == params.Base58Prefix(CChainParams::BLS_PUBKEY_ADDRESS);
+    if (!fKnownVersion)
+        return false;
```
all changes in base58.h/.cpp have vanished, so I'm going to resolve all comments for these

codablock commented on src/crypto/bls.cpp:
```diff
@@ -0,0 +1,797 @@
+    }
+#endif
+    return true;
+}
```
Code has vanished

codablock commented on src/crypto/bls.h:
```diff
@@ -0,0 +1,394 @@
+            memset(buf, 0, size);
+        } else {
+            bool ok = InternalGetBuf(buf, size);
+            assert(ok);
```
Never put important logic into the `assert(...)`. You never know if the assert will stay in release builds...
It does in our code, because it's forced somewhere...but nothing we should rely on.

codablock commented:
Also updated the example dash.conf to work with the changes we did in develop

codablock commented:
Closing this PR as the code has been merged through smaller PRs.
